### PR TITLE
Merge branch uq surrogates into master

### PR DESCRIPTION
**Pull request by MichaelReh**
The main new features added with this PR are
B-spline bases NakBsplineExtended and NakPBspline
The B-spline Surrogate convenience classes SplineResponseSurface and SplineResponseSurfaceVector
The sgpp::base::Distribution class

### CODE REVIEW

File: optimization/tests/testActiveSubspaces.cpp [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+//  sgpp::base::DataMatrix evaluationPoints = ASM.getEvaluationPoints();
+//  sgpp::base::DataVector functionValues = ASM.getFunctionValues();
+//  size_t responseLevel = 3;
+//  responseSurf.createRegularReducedSurfaceFromData(evaluationPoints, functionValues,
+//  responseLevel); sgpp::base::DataVector v(numDim, 0.3371); double objectiveFunctionEval =
+//  objectiveFunctionResponseSurface(v); double responseSurfEval = responseSurf.eval(v); double
+//  epsilon = 1e-15; BOOST_CHECK_SMALL(fabs(objectiveFunctionEval - responseSurfEval), epsilon);
+//}
+//
+// BOOST_AUTO_TEST_SUITE_END()
```
Comment (leiterrl) [inline optimization/tests/testActiveSubspaces.cpp:274]: Is this file relevant?

**Review by leiterrl — CHANGES_REQUESTED** [review]
Added some minor comments. Also, some files are missing copyright notices. There should be a script in /tools which can identify all of those. Other than that the PR looks good to merge!

File: datadriven/python/uq/refinement/RefinementStrategy.py [hunk 2]
```diff
@@ -434,6 +437,79 @@ def update(self, grid, v, gpi, params, *args, **kws):
...
+    """
+    This is pm1d specific! DO NOT MERGE INTO MASTER!
+    Special Refinement criterion for the pm1d CO2 storage example
+    It replicates the WeightedL2OptRanking but calculates the secondMoment with Monte Carlo
+    instead of using estimated densities.
+    This is used in particular for calculating results of the pm1d code with B-splines as currently
+    there are no estimated densities available for them.
+    The Monte Carlo guess is based on 10000 sample points given by the authors of the pm1d code
+    As  the basis function integrals are independent of the actual setting they are precalculated and saved in files.
+    """
```
Comment (leiterrl) [inline datadriven/python/uq/refinement/RefinementStrategy.py:451]: This comment is a bit worrying :sweat_smile:

File: base/src/sgpp/base/tools/sle/solver/Eigen.cpp [hunk 3]
```diff
@@ -76,7 +76,11 @@ bool Eigen::solve(SLE& system, DataMatrix& B, DataMatrix& X) const {
   size_t rowsDone = 0;
 
 // parallelize only if the system is cloneable
-#pragma omp parallel if (system.isCloneable()) shared(system, A, nnz, rowsDone) default(none)
+// ToDo (rehmemk): On my Ubuntu 19.10 instance the following did not work anymore
+// and I replaced it by a simple default(shared). Does this introduce problems?
+// (same in Armadillo.cpp)
+#pragma omp parallel if (system.isCloneable()) shared(system, A, nnz, rowsDone, n)  // default(none)
+  // #pragma omp parallel if (system.isCloneable()) default(shared)
```
[the remark this answers is not in the document]
Comment (MichaelReh) [reply base/src/sgpp/base/tools/sle/solver/Eigen.cpp:83]: We discussed that this is fine some time back. So I#ll just remove the comment

File: datadriven/examples/learnerRegressionTest.cpp [hunk 4]
```diff
@@ -128,15 +131,15 @@ std::vector<sgpp::datadriven::RegularizationConfiguration> getConfigs() {
...
-    }
+    //    {
+    //      // Laplace
+    //      const auto regularizationType = sgpp::datadriven::RegularizationType::Laplace;
+    //      auto regularizationConfig = sgpp::datadriven::RegularizationConfiguration();
+    //      regularizationConfig.type_ = regularizationType;
+    //      regularizationConfig.lambda_ = lambda;
+    //      regularizationConfig.exponentBase_ = 0.25;
+    //      result.push_back(regularizationConfig);
+    //    }
```
[the remark this answers is not in the document]
Comment (MichaelReh) [reply datadriven/examples/learnerRegressionTest.cpp:142]: No and I can't remember why I commented this. I uncommented it and the tests run fine.

File: optimization/tests/testActiveSubspaces.cpp (hunk 1 above)
Comment (MichaelReh) [reply optimization/tests/testActiveSubspaces.cpp:274]: No, removed.

File: datadriven/python/uq/refinement/RefinementStrategy.py (hunk 2 above)
Comment (MichaelReh) [reply datadriven/python/uq/refinement/RefinementStrategy.py:451]: Thanks for noticing. Removed

File: base/src/sgpp_base.hpp [hunk 5]
```diff
@@ -115,7 +117,10 @@
 #include <sgpp/base/grid/generation/functors/SurplusRefinementFunctor.hpp>
 #include <sgpp/base/grid/generation/functors/SurplusVolumeCoarseningFunctor.hpp>
 #include <sgpp/base/grid/generation/functors/SurplusVolumeRefinementFunctor.hpp>
-
+/*#include <sgpp/base/grid/generation/functors/WeightedErrorRefinementFunctor.hpp>
+#include <sgpp/base/grid/generation/functors/ClassificationRefinementFunctor.hpp>
+#include <sgpp/base/grid/generation/functors/PersistentErrorRefinementFunctor.hpp>
+#include <sgpp/base/grid/generation/functors/PredictiveRefinementDimensionIndicator.hpp>*/
```
[the remark this answers is not in the document]
Comment (MichaelReh) [reply base/src/sgpp_base.hpp:123]: removed

**Review by leiterrl — APPROVED** [review]

**Review by leiterrl — APPROVED** [review]

**Pull request merged**